File: 프로그래머스/1/42889. 실패율/실패율.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

double arr[502];
double k[502];

vector<pair<double,int>>v;

bool compare(pair<double,int> a, pair<double,int> b){
    if(a.first != b.first){
        return a.first > b.first;
    }
    return a.second < b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    
    for(int i=0;i<stages.size();i++){
        arr[stages[i]]++;
        for(int j=1;j<=stages[i];j++){
            k[j]++;
        }
    }
    
    for(int i=1;i<=N;i++){
        if(arr[i]==0 || k[i]==0) {
            v.push_back({0.0,i});
        }
        else v.push_back({(arr[i]/k[i]), i});
    }
    
    sort(v.begin(),v.end(),compare);
    for(int i=0;i<v.size();i++){
        answer.push_back(v[i].second);
    }
    
    return answer;
}

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;


class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = {};
        
        int[] ch = new int[N+2];
        
        for(int stage : stages){
            ch[stage] += 1;
        }
        
        HashMap<Integer, Double> fails = new HashMap<>();
        
        double total = stages.length;
        
        for(int i=1;i<=N;i++){
            if(ch[i]==0){
                fails.put(i,0.);
            }else{
                fails.put(i, ch[i] / total);
                total -= ch[i];
            }
        }
        
       return fails.entrySet().stream().sorted((o1,o2)->o1.getValue().equals(o2.getValue()) ? 
             Integer.compare(o1.getKey(), o2.getKey()) : Double.compare(o2.getValue(), o1.getValue())).mapToInt(HashMap.Entry::getKey).toArray();
    }
}